h2o-core/src/main/java/water/AutoBuffer.java

```diff
@@ -128,17 +128,22 @@
     _persist = 0;               // No persistance
   }
 
-  /** Incoming TCP request.  Make a read-mode AutoBuffer from the open Channel,
-   *  figure the originating H2ONode from the first few bytes read. */
-  AutoBuffer( SocketChannel sock ) throws IOException {
+  /** Incoming TCP request.  Make a read-mode AutoBuffer from the open Channel.
+   * If isSenderH2ONode is set to true than also figure the originating H2ONode from the first few bytes read, otherwise
+   * just don't set _h2o field because the communication originates from non-h2o node such as spark executor*/
+  AutoBuffer( SocketChannel sock, boolean isSenderH2ONode ) throws IOException {
     _chan = sock;
     raisePriority();            // Make TCP priority high
     _bb = BBP_BIG.make();       // Get a big / TPC-sized ByteBuffer
     _bb.flip();
     _read = true;               // Reading by default
     _firstPage = true;
     // Read Inet from socket, port from the stream, figure out H2ONode
-    _h2o = H2ONode.intern(sock.socket().getInetAddress(), getPort());
+    if(isSenderH2ONode) {
```

Comment:
On spark side I create the sockets using H2O's Autobuffer similarly to 
```
val ab = new AutoBuffer()
     ab.put1(ExternalFrameHandler.INIT_BYTE)
     ab.putInt(ExternalFrameHandler.DOWNLOAD_FRAME)
     ab.putStr(keyName)
     ab.putA1(types)
     ab.putInt(chunkIdx)
```

@mdymczyk I'm not exactly sure how the ssl implementation looks like at this moment - do you need to create pair of keys between each communicating node ? If yes, we could just generate one key for spark driver and pass it from there to all discovered executors.